language-plutus-core/untyped-plutus-core-test/Evaluation/ApplyBuiltinName.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications      #-}
+
+module Evaluation.ApplyBuiltinName where
+
+import           PlutusPrelude
+
+import           Language.UntypedPlutusCore
+
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
+import           Language.PlutusCore.Evaluation.Machine.ExBudgetingDefaults
+import           Language.PlutusCore.Evaluation.Machine.Exception
+import           Language.PlutusCore.Generators
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Universe
+
+import           Control.Monad.Except
+import qualified Data.ByteString.Lazy                                       as BSL
+import qualified Data.ByteString.Lazy.Hash                                  as Hash
+import           Data.Proxy
+import           Hedgehog
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+-- | A simplified (because we don't need the full generality here) version of
+-- 'Language.PlutusCore.Generators.Internal.Entity.genIterAppValue'.
+genArgsRes
+    :: Generatable uni
+    => TypeScheme (Term Name uni ()) as res -> FoldArgs as res -> Gen ([Term Name uni ()], res)
+genArgsRes (TypeSchemeResult _)       y = return ([], y)
+genArgsRes (TypeSchemeArrow _ schB)   f = do
+    TermOf v x <- genTypedBuiltinDef AsKnownType
+    first (v :) <$> genArgsRes schB (f x)
+genArgsRes (TypeSchemeAllType _ schK) f = genArgsRes (schK Proxy) f
+
+type AppErr = EvaluationException () () (Term Name DefaultUni ())
+newtype AppM a = AppM
+    { unAppM :: Either AppErr a
+    } deriving newtype (Functor, Applicative, Monad, MonadError AppErr)
+
+instance SpendBudget AppM (Term Name DefaultUni ()) where
+    spendBudget _ _ _ = pure ()
+    builtinCostParams = pure defaultCostModel
+
+-- | This shows that the builtin application machinery accepts untyped terms.
+prop_applyBuiltinName
+    :: (uni ~ DefaultUni, KnownType (Term Name uni ()) res)
+    => TypedBuiltinName (Term Name uni ()) args res
+       -- ^ A (typed) builtin name to apply.
+    -> FoldArgs args res
+       -- ^ The semantics of the builtin name. E.g. the semantics of
+       -- 'AddInteger' (and hence 'typedAddInteger') is '(+)'.
+    -> Property
+prop_applyBuiltinName (TypedBuiltinName name sch) op = property $ do
+    (args, res) <- forAllNoShow $ genArgsRes sch op
+    let rhs = evaluationConstAppResult $ makeKnown res
+    case unAppM $ applyBuiltinName name args of
+        Left _    -> fail $ "Failure while checking an application of " ++ show name
+        Right lhs -> lhs === rhs
+
+test_typedAddInteger :: TestTree
+test_typedAddInteger
+    = testProperty "typedAddInteger"
+    $ prop_applyBuiltinName typedAddInteger (+)
```

Comment:
It would be nice to just have a function that gives you both the typed version of a builtin and its denotation, so that we don't need to replicate all this baggage here. I'll look into it.